From 081829674bd17c3af2e37e55d8acda8e8f117d23 Mon Sep 17 00:00:00 2001
From: Nikita Yushchenko <nikita.yoush@cogentembedded.com>
Date: Sun, 14 May 2023 12:22:37 +0600
Subject: [PATCH 17/92] arm64: dts: renesas: add device trees for vc4

r8a779f0-vc4.dts - for board schematics as of 2022-12-09
r8a779f0-vc4V1.dts - for board schematics as of 2021-07-13

*-ctrl-domain.dts - versions to use when control domain registers are
accessible for Linux.

Signed-off-by: Nikita Yushchenko <nikita.yoush@cogentembedded.com>
---
 arch/arm64/boot/dts/renesas/Makefile          |   2 +
 .../dts/renesas/r8a779f0-vc4-ctrl-domain.dts  |  13 +
 arch/arm64/boot/dts/renesas/r8a779f0-vc4.dts  | 460 ++++++++++++++++++
 .../renesas/r8a779f0-vc4V1-ctrl-domain.dts    |  75 +++
 .../arm64/boot/dts/renesas/r8a779f0-vc4V1.dts |  49 ++
 5 files changed, 599 insertions(+)
 create mode 100644 arch/arm64/boot/dts/renesas/r8a779f0-vc4-ctrl-domain.dts
 create mode 100644 arch/arm64/boot/dts/renesas/r8a779f0-vc4.dts
 create mode 100644 arch/arm64/boot/dts/renesas/r8a779f0-vc4V1-ctrl-domain.dts
 create mode 100644 arch/arm64/boot/dts/renesas/r8a779f0-vc4V1.dts

diff --git a/arch/arm64/boot/dts/renesas/Makefile b/arch/arm64/boot/dts/renesas/Makefile
index 1499361d6ecf..826e30aed3a1 100644
--- a/arch/arm64/boot/dts/renesas/Makefile
+++ b/arch/arm64/boot/dts/renesas/Makefile
@@ -74,6 +74,8 @@ dtb-$(CONFIG_ARCH_R8A779A0) += r8a779a0-falcon.dtb
 dtb-$(CONFIG_ARCH_R8A779G0) += r8a779g0-whitehawk.dtb
 
 dtb-$(CONFIG_ARCH_R8A779F0) += r8a779f0-spider.dtb
+dtb-$(CONFIG_ARCH_R8A779F0) += r8a779f0-vc4.dtb r8a779f0-vc4-ctrl-domain.dtb
+dtb-$(CONFIG_ARCH_R8A779F0) += r8a779f0-vc4V1.dtb r8a779f0-vc4V1-ctrl-domain.dtb
 
 dtb-$(CONFIG_ARCH_R8A77970) += r8a77970-eagle-function.dtb
 dtb-$(CONFIG_ARCH_R8A77970) += r8a77970-es1-eagle.dtb r8a77970-es1-eagle-function.dtb
diff --git a/arch/arm64/boot/dts/renesas/r8a779f0-vc4-ctrl-domain.dts b/arch/arm64/boot/dts/renesas/r8a779f0-vc4-ctrl-domain.dts
new file mode 100644
index 000000000000..d158c99dfe69
--- /dev/null
+++ b/arch/arm64/boot/dts/renesas/r8a779f0-vc4-ctrl-domain.dts
@@ -0,0 +1,13 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Device Tree Source for the VC4 vehicle computer board
+ *
+ * This dts version requires full access for CA55 cores (application domain)
+ * to the control domain. Means G4 processor domain needs to grant access
+ * before Linux is booted.
+ *
+ * Copyright (C) 2021-2023 Renesas Electronics Corp.
+ */
+
+#include "r8a779f0-vc4.dts"
+#include "r8a779f0-ctrl-domain.dtsi"
diff --git a/arch/arm64/boot/dts/renesas/r8a779f0-vc4.dts b/arch/arm64/boot/dts/renesas/r8a779f0-vc4.dts
new file mode 100644
index 000000000000..bfefa7f35846
--- /dev/null
+++ b/arch/arm64/boot/dts/renesas/r8a779f0-vc4.dts
@@ -0,0 +1,460 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Device Tree Source for the VC4 vehicle computer board
+ *
+ * Copyright (C) 2021-2023 Renesas Electronics Corp.
+ */
+
+/dts-v1/;
+
+#include "r8a779f0.dtsi"
+#include <dt-bindings/gpio/gpio.h>
+
+/ {
+	model = "Renesas VC4 board based on r8a779f0";
+	compatible = "renesas,vc4", "renesas,r8a779f0";
+
+	aliases {
+		serial0 = &scif3;
+		serial1 = &scif0;
+		/* serial0 = &hscif0 */
+		/* serial1 = &hscif1 */
+		eth0 = &rswitch;
+	};
+
+	chosen {
+		stdout-path = "serial0:115200n8";
+	};
+
+	memory@48000000 {
+		device_type = "memory";
+		/* First 128MB is reserved for secure area. */
+		/* Last 512MB is reserved for CR. */
+		reg = <0x0 0x48000000 0x0 0x58000000>;
+	};
+
+	memory@480000000 {
+		device_type = "memory";
+		reg = <0x4 0x80000000 0x0 0x80000000>;
+	};
+
+	reg_1p8v: regulator-1p8v {
+		compatible = "regulator-fixed";
+		regulator-name = "fixed-1.8V";
+		regulator-min-microvolt = <1800000>;
+		regulator-max-microvolt = <1800000>;
+		regulator-boot-on;
+		regulator-always-on;
+	};
+
+	reg_3p3v: regulator-3p3v {
+		compatible = "regulator-fixed";
+		regulator-name = "fixed-3.3V";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		regulator-boot-on;
+		regulator-always-on;
+	};
+
+	vcc_sdhi: regulator-vcc-sdhi {
+		compatible = "regulator-fixed";
+
+		regulator-name = "SDHI Vcc";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+
+		gpio = <&gpio1 24 GPIO_ACTIVE_HIGH>;
+		enable-active-high;
+	};
+
+	vccq_sdhi: regulator-vccq-sdhi {
+		compatible = "regulator-gpio";
+
+		regulator-name = "SDHI VccQ";
+		regulator-min-microvolt = <1800000>;
+		regulator-max-microvolt = <3300000>;
+
+		gpio = <&gpio2 16 GPIO_ACTIVE_HIGH>;
+		gpios-states = <1>;
+		states = <3300000 1>, <1800000 0>;
+	};
+};
+
+&extal_clk {
+	clock-frequency = <20000000>;
+};
+
+&extalr_clk {
+	clock-frequency = <32768>;
+};
+
+&gpio3 {
+	tsn0-pre-mux {
+		gpio-hog;
+		gpios = <8 GPIO_ACTIVE_HIGH>;
+		output-high;
+		line-name = "TSN0-PRE-MUX";
+	};
+
+	tsn0_mux: tsn0-mux {
+		gpio-hog;
+		gpios = <6 GPIO_ACTIVE_HIGH>;
+		/* For tsn0_phy_2g5 */
+		/* output-low; */
+		/* For tsn0_phy_1g */
+		output-high;
+		line-name = "TSN0-MUX";
+	};
+
+	tsn1_mux: tsn1-mux {
+		gpio-hog;
+		gpios = <14 GPIO_ACTIVE_HIGH>;
+		/* For tsn1_phy_2g5 */
+		/* output-low; */
+		/* For tsn1_phy_1g */
+		output-high;
+		line-name = "TSN0-MUX";
+	};
+
+	tsn2-mux {
+		gpio-hog;
+		gpios = <7 GPIO_ACTIVE_HIGH>;
+		/* For tsn2_phy_1g_rh */
+		/* output-low; */
+		/* For tsn2_phy_1g */
+		output-high;
+		line-name = "TSN2-MUX";
+	};
+};
+
+&i2c1 {
+	status = "okay";
+
+	pinctrl-0 = <&i2c1_pins>;
+	pinctrl-names = "default";
+
+	clock-frequency = <400000>;
+
+	rtc@6f {
+		compatible = "isil,isl1208";
+		reg = <0x6f>;
+	};
+};
+
+&i2c4 {
+	status = "okay";
+
+	pinctrl-0 = <&i2c4_pins>;
+	pinctrl-names = "default";
+
+	clock-frequency = <400000>;
+
+	eeprom@50 {
+		compatible = "st,24c16", "atmel,24c16";
+		reg = <0x50>;
+
+		vcc-supply = <&reg_3p3v>;
+
+		label = "board-data";
+		pagesize = <8>;
+	};
+};
+
+&hscif0 {
+	/* Please use exclusively to the scif3 node */
+	status = "disabled";
+
+	pinctrl-0 = <&hscif0_pins>;
+	pinctrl-names = "default";
+
+	uart-has-rtscts;
+};
+
+&hscif1 {
+	/* Please use exclusively to the scif0 node */
+	status = "disabled";
+
+	pinctrl-0 = <&hscif1_pins>;
+	pinctrl-names = "default";
+
+	uart-has-rtscts;
+};
+
+&mmc0 {
+	status = "okay";
+
+	pinctrl-0 = <&sd_pins>;
+	pinctrl-1 = <&sd_pins>;
+	pinctrl-names = "default", "state_uhs";
+
+	vmmc-supply = <&vcc_sdhi>;
+	vqmmc-supply = <&vcc_sdhi>;
+	cd-gpios = <&gpio1 23 GPIO_ACTIVE_LOW>;
+	bus-width = <4>;
+};
+
+&pfc {
+	i2c1_pins: i2c1 {
+		groups = "i2c1";
+		function = "i2c1";
+	};
+
+	i2c4_pins: i2c4 {
+		groups = "i2c4";
+		function = "i2c4";
+	};
+
+	hscif0_pins: hscif0 {
+		groups = "hscif0_data", "hscif0_ctrl";
+		function = "hscif0";
+	};
+
+	hscif1_pins: hscif1 {
+		groups = "hscif1_data", "hscif1_ctrl";
+		function = "hscif1";
+	};
+
+	pcie0_pins: pcie0 {
+		groups = "pcie0_clkreq_n";
+		function = "pcie";
+	};
+
+	rpc_pins: rpc {
+		qspi0 {
+			groups = "qspi0_data4", "qspi0_ctrl";
+			function = "qspi0";
+		};
+		qspi1 {
+			groups = "qspi1_data4", "qspi1_ctrl";
+			function = "qspi1";
+		};
+	};
+
+	rswitch_pins: rswitch {
+		tsn0_mdio {
+			groups = "tsn0_mdio";
+			function = "tsn0";
+		};
+
+		tsn1_mdio {
+			groups = "tsn1_mdio";
+			function = "tsn1";
+		};
+
+		tsn2_mdio {
+			groups = "tsn2_mdio";
+			function = "tsn2";
+		};
+	};
+
+	scif0_pins: scif0 {
+		groups = "scif0_data", "scif0_ctrl";
+		function = "scif0";
+	};
+
+	scif3_pins: scif3 {
+		groups = "scif3_data", "scif3_ctrl";
+		function = "scif3";
+	};
+
+	scif_clk_pins: scif_clk {
+		groups = "scif_clk";
+		function = "scif_clk";
+	};
+
+	sd_pins: sd {
+		groups = "mmc_data4", "mmc_ctrl";
+		function = "mmc";
+		power-source = <3300>;
+	};
+};
+
+&pciec0 {
+	status = "okay";
+
+	pinctrl-0 = <&pcie0_pins>;
+	pinctrl-names = "default";
+};
+
+&pciec1 {
+	// Previously, enabling this caused a hang
+	status = "disabled";
+};
+
+&pcie_bus_clk {
+	clock-frequency = <100000000>;
+};
+
+&rpc {
+	status = "okay";
+
+	pinctrl-0 = <&rpc_pins>;
+	pinctrl-names = "default";
+
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <133000000>;
+		spi-rx-bus-width = <4>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			bootparam@0 {
+				reg = <0x00000000 0x00040000>;
+				read-only;
+			};
+			icumxa_loader@40000 {
+				reg = <0x00040000 0x00200000>;
+				read-only;
+			};
+			cert_header@240000 {
+				reg = <0x00240000 0x00040000>;
+				read-only;
+			};
+			secure_fw@280000 {
+				reg = <0x00280000 0x00100000>;
+				read-only;
+			};
+			icumh_payload@280000 {
+				reg = <0x00380000 0x00100000>;
+				read-only;
+			};
+			secure_monitor@480000 {
+				reg = <0x00480000 0x00080000>;
+				read-only;
+			};
+			rtos@500000 {
+				reg = <0x00500000 0x00400000>;
+				read-only;
+			};
+			g4mh_payload@900000 {
+				reg = <0x00900000 0x00100000>;
+				read-only;
+			};
+			space1@a00000 {
+				reg = <0x00a00000 0x00300000>;
+			};
+			uboot-env@d00000 {
+				reg = <0x00d00000 0x00040000>;
+			};
+			space2@d40000 {
+				reg = <0x00d40000 0x000c0000>;
+			};
+			uboot@e00000 {
+				reg = <0x00e00000 0x00200000>;
+				read-only;
+			};
+			tee@1000000 {
+				reg = <0x01000000 0x00100000>;
+				read-only;
+			};
+			space3@1100000 {
+				reg = <0x01100000 0x02f00000>;
+			};
+		};
+	};
+};
+
+&rswitch {
+	status = "okay";
+
+	pinctrl-0 = <&rswitch_pins>;
+	pinctrl-names = "default";
+
+	ports {
+		#address-cells = <1>;
+		#size-cells = <0>;
+		port@0 {
+			reg = <0>;
+			/* phy-mode = "usxgmii"; */
+			/* phy-handle = <&tsn0_phy_2g5>; */
+			phy-mode = "sgmii";
+			phy-handle = <&tsn0_phy_1g>;
+			#address-cells = <1>;
+			#size-cells = <0>;
+			tsn0_phy_2g5: ethernet-phy@0 {
+				reg = <0>;
+				compatible = "ethernet-phy-ieee802.3-c45";
+				status = "disabled";
+			};
+			tsn0_phy_1g: ethernet-phy@3 {
+				reg = <3>;
+				compatible = "ethernet-phy-ieee802.3-c45";
+				status = "okay";
+			};
+		};
+		port@1 {
+			reg = <1>;
+			/* phy-mode = "usxgmii"; */
+			/* phy-handle = <&tsn1_phy_2g5>; */
+			phy-mode = "sgmii";
+			phy-handle = <&tsn1_phy_1g>;
+			#address-cells = <1>;
+			#size-cells = <0>;
+			tsn1_phy_2g5: ethernet-phy@0 {
+				reg = <0>;
+				compatible = "ethernet-phy-ieee802.3-c45";
+				status = "disabled";
+			};
+			tsn1_phy_1g: ethernet-phy@3 {
+				reg = <3>;
+				compatible = "ethernet-phy-ieee802.3-c45";
+				status = "okay";
+			};
+		};
+		port@2 {
+			reg = <2>;
+			phy-mode = "sgmii";
+			/* phy-handle = <&tsn2_phy_1g_rh>; */
+			phy-handle = <&tsn2_phy_1g>;
+			#address-cells = <1>;
+			#size-cells = <0>;
+			tsn2_phy_1g_rh: ethernet-phy@1 {
+				reg = <1>;
+				compatible = "ethernet-phy-ieee802.3-c45";
+				status = "disabled";
+			};
+			tsn2_phy_1g: ethernet-phy@3 {
+				reg = <3>;
+				compatible = "ethernet-phy-ieee802.3-c45";
+				status = "okay";
+			};
+		};
+	};
+};
+
+&rwdt {
+	status = "disabled";
+
+	timeout-sec = <60>;
+};
+
+&scif0 {
+	/* Please use exclusively to the hscif1 node */
+	status = "okay";
+
+	pinctrl-0 = <&scif0_pins>;
+	pinctrl-names = "default";
+
+	uart-has-rtscts;
+};
+
+&scif3 {
+	/* Please use exclusively to the hscif0 node */
+	status = "okay";
+
+	pinctrl-0 = <&scif3_pins>;
+	pinctrl-names = "default";
+
+	uart-has-rtscts;
+};
+
+&scif_clk {
+	pinctrl-0 = <&scif_clk_pins>;
+	pinctrl-names = "default";
+
+	clock-frequency = <24000000>;
+};
diff --git a/arch/arm64/boot/dts/renesas/r8a779f0-vc4V1-ctrl-domain.dts b/arch/arm64/boot/dts/renesas/r8a779f0-vc4V1-ctrl-domain.dts
new file mode 100644
index 000000000000..23a55378cc08
--- /dev/null
+++ b/arch/arm64/boot/dts/renesas/r8a779f0-vc4V1-ctrl-domain.dts
@@ -0,0 +1,75 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Device Tree Source for the VC4 vehicle computer board
+ * For V1 schematics, as of 2021-07-13
+ *
+ * This dts version requires full access for CA55 cores (application domain)
+ * to the control domain. Means G4 processor domain needs to grant access
+ * before Linux is booted.
+ *
+ * Copyright (C) 2021-2023 Renesas Electronics Corp.
+ */
+
+#include "r8a779f0-vc4V1.dts"
+#include "r8a779f0-ctrl-domain.dtsi"
+
+/* SDHI_PWR_SEL is at GPIO4_05 */
+
+&vccq_sdhi {
+	status = "okay";
+	gpio = <&gpio4 5 GPIO_ACTIVE_HIGH>;
+};
+
+&mmc0 {
+	vqmcc-supply = <&vccq_sdhi>;
+};
+
+/* TSN0_SEL_1_25 is at GPIO4_06 */
+
+&gpio4 {
+	tsn0-mux {
+		gpio-hog;
+		gpios = <6 GPIO_ACTIVE_HIGH>;
+		/* For tsn0_phy_2g5 */
+		/* output-low; */
+		/* For tsn0_phy_1g */
+		output-high;
+		line-name = "TSN0-MUX";
+	};
+};
+
+/* TSN1_SEL_1_25 is at GPIO4_07 */
+
+&gpio4 {
+	tsn1-mux {
+		gpio-hog;
+		gpios = <7 GPIO_ACTIVE_HIGH>;
+		/* For tsn1_phy_2g5 */
+		/* output-low; */
+		/* For tsn1_phy_1g */
+		output-high;
+		line-name = "TSN1-MUX";
+	};
+};
+
+/* TSN0_WAKE is at GPIO6_22. When using tsn0_phy_2g5, keep high */
+
+&gpio6 {
+	/* tsn0-no-wake {
+		gpio-hog;
+		gpios = <22 GPIO_ACTIVE_HIGH>;
+		output-high;
+		line-name = "TSN0-NO-WAKE";
+	}; */
+};
+
+/* TSN1_WAKE is at GPIO6_21. When using tsn1_phy_2g5, keep high */
+
+&gpio6 {
+	/* tsn1-no-wake {
+		gpio-hog;
+		gpios = <21 GPIO_ACTIVE_HIGH>;
+		output-high;
+		line-name = "TSN1-NO-WAKE";
+	}; */
+};
diff --git a/arch/arm64/boot/dts/renesas/r8a779f0-vc4V1.dts b/arch/arm64/boot/dts/renesas/r8a779f0-vc4V1.dts
new file mode 100644
index 000000000000..9536a3ad1e62
--- /dev/null
+++ b/arch/arm64/boot/dts/renesas/r8a779f0-vc4V1.dts
@@ -0,0 +1,49 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Device Tree Source for the VC4 vehicle computer board
+ * For V1 schematics, as of 2021-07-13
+ *
+ * Copyright (C) 2021-2023 Renesas Electronics Corp.
+ */
+
+#include "r8a779f0-vc4.dts"
+
+/* SDHI_PWR_SEL is at GPIO4_05 - n/a without control domain access */
+
+&vccq_sdhi {
+	status = "disabled";
+};
+
+&mmc0 {
+	vqmcc-supply = <&vcc_sdhi>;
+};
+
+/* TSN0_SEL_1_25 is at GPIO4_06 - n/a without control domain access */
+
+/delete-node/ &tsn0_mux;
+
+/* TSN1_SEL_1_25 is at GPIO4_07 - n/a without control domain access */
+
+/delete-node/ &tsn1_mux;
+
+/* TSN0_SLEEP is at GPIO3_15. When using tsn0_phy_2g5, keep high */
+
+&gpio3 {
+	/* tsn0-no-sleep {
+		gpio-hog;
+		gpios = <15 GPIO_ACTIVE_HIGH>;
+		output-high;
+		line-name = "TSN0-NO-SLEEP";
+	}; */
+};
+
+/* TSN1_SLEEP is at GPIO3_12. When using tsn1_phy_2g5, keep high */
+
+&gpio3 {
+	/* tsn0-no-sleep {
+		gpio-hog;
+		gpios = <12 GPIO_ACTIVE_HIGH>;
+		output-high;
+		line-name = "TSN1-NO-SLEEP";
+	}; */
+};
-- 
2.30.2

